pyqtorch/primitive.py

```diff
@@ -138,164 +58,154 @@ def eigenvals_generator(self) -> Tensor:
         Returns:
             Eigenvalues of the generator operator.
         """
-        if self.pauli_generator is not None:
-            return torch.linalg.eigvalsh(self.pauli_generator).reshape(-1, 1)
+        if self.generator is not None:
+            return torch.linalg.eigvalsh(self.generator).reshape(-1, 1)
         pass
 
-    @cached_property
-    def spectral_gap(self) -> Tensor:
-        """Difference between the moduli of the two largest eigenvalues of the generator.
 
-        Returns:
-            Tensor: Spectral gap value.
-        """
-        spectrum = self.eigenvals_generator
-        spectral_gap = torch.unique(torch.abs(torch.tril(spectrum - spectrum.T)))
-        return spectral_gap[spectral_gap.nonzero()]
+class ControlledPrimitive(Primitive):
+    """Primitive applied depending on control qubits.
+
+    Attributes:
+        operation (Tensor): Unitary tensor U.
+        control (int | tuple[int, ...]): List of qubits acting as controls.
+        target (int | tuple[int, ...]): List of qubits operations acts on.
+    """
 
-    def tensor(
+    def __init__(
         self,
-        values: dict[str, Tensor] = {},
-        embedding: Embedding | None = None,
-        full_support: tuple[int, ...] | None = None,
-        diagonal: bool = False,
-    ) -> Tensor:
-        if diagonal:
-            raise NotImplementedError
-        blockmat = self.unitary(values, embedding)
-        if full_support is None:
-            return blockmat
-        else:
-            return expand_operator(blockmat, self.qubit_support, full_support)
+        operation: str | Tensor,
+        control: int | tuple[int, ...],
+        target: int | tuple[int, ...],
+    ):
+        support = Support(target, control)
+        if isinstance(operation, str):
+            operation = OPERATIONS_DICT[operation]
+        operation = _controlled(
+            unitary=operation.unsqueeze(2),
```

Comment:
Since we're at it, maybe change the `_controlled` function to `controlled` and call the `unitary` argument `operation` ?